finance_dl/schwab.py

```diff
@@ -0,0 +1,303 @@
+                if sel:
+                    acct = self.get_account_from_container(sel[0])
+                    if acct is not None:
+                        return acct.number == account.number
+                return False
+
+            WebDriverWait(self.driver, 30).until(
+                EC.invisibility_of_element_located(
+                    (By.CSS_SELECTOR, "sdps-account-selector__header")))
+            # Make sure Schwab registers the selection of the new account.
+            time.sleep(1)
+
+            return account
+        return None
+
+    def get_account_links(self) -> List[Any]:
+        selector_button, = self.get_elements_wait("button.account-selector-button")
+        selector_button.click()
+
+        return self.get_elements_wait(
+            "li.sdps-account-selector__list-item a"
+        )
+
+    def get_account_from_container(self, container: Any) -> Optional[Account]:
+        spans = container.find_elements(By.CSS_SELECTOR, "span")
+        if len(spans) < 2:
+            return None
+        label_span, number_span = spans[:2]
+        number = number_span.text
+        if not number:
+            return None
+        return Account(label=label_span.text, number=number)
+
+    def login_if_needed(self) -> None:
+        login_frames = []
+        logout_buttons = []
+
+        def predicate(driver):
+            login_frames.extend(
+                driver.find_elements(By.CSS_SELECTOR, "iframe#lmsSecondaryLogin")
+            )
+            logout_buttons.extend(
+                driver.find_elements(By.CSS_SELECTOR, "button.logout")
+            )
+            return login_frames or logout_buttons
+
+        WebDriverWait(self.driver, 30).until(
+            predicate, message="Did not find either app or login page."
+        )
+
+        if logout_buttons:
+            logger.info("Already logged in.")
+            return
+
+        (login_frame,) = login_frames
+
+        logger.info("Logging in.")
+
+        self.driver.switch_to.frame(login_frame)
+
+        (username,) = self.get_elements("input#LoginId")
+
+        username.send_keys(self.credentials["username"])
+
+        logger.info("Looking for password field.")
+        (password,) = self.get_elements("input#Password")
+        password.send_keys(self.credentials["password"])
+
+        password.send_keys(Keys.ENTER)
+
+        logger.info("Waiting for app to load.")
+
+        self.driver.switch_to.default_content()
+        # Just confirm we are logged in.
+        self.get_elements_wait("button.logout")
+
+        logger.info("Logged in.")
+        return
+
+    def get_account_dirs(self, account: Account) -> Tuple[str, str]:
+        acct_dir = os.path.join(self.output_directory, account.number)
+        pos_dir = os.path.join(acct_dir, "positions")
+        if not os.path.exists(pos_dir):
+            os.makedirs(pos_dir)
+        return acct_dir, pos_dir
+
+    def get_last_fetched_date(self, account_dir: str) -> Optional[datetime.date]:
```

Comment:
Note: You might find you can relax the zero-overlap constraint if you merge your downloaded csv files using the `finance_dl.csv_merge` module.  That handles duplicate rows as follows: if you merge files A and B, and A contains 1 copy of a row, and B contains 2 copies of the same row, the merged file contains 2 copies of that row.

Reply:
I looked at this, but the importer relies on the precise format of Schwab CSV files, which is CSV plus extra non-CSV lines with used information (account names etc); adapting the merge to account for this feels a bit fiddly and it seems simpler to keep and use the files exactly as received from Schwab. I think the downloader can maintain zero overlap with correctness pretty easily, by never downloading a current partial day.